api/v1alpha1/mysqlaaduser_conversion.go

```diff
@@ -4,30 +4,55 @@
 package v1alpha1
 
 import (
-	"github.com/pkg/errors"
+	"sort"
+
 	"sigs.k8s.io/controller-runtime/pkg/conversion"
 
 	"github.com/Azure/azure-service-operator/api/v1alpha2"
 )
 
 var _ conversion.Convertible = &MySQLAADUser{}
 
+// To avoid losing information converting a v1alpha2 instance into
+// v1alpha1, we stash the affected fields (roles and database roles)
+// into a json-serialised struct in the annotations. When converting
+// back to a v1alpha2 instance we use any stashed values, layering any
+// changes to the database roles back over the top. The conversions
+// will only return errors if JSON marshalling/unmarshalling fails.
+
 func (src *MySQLAADUser) ConvertTo(dstRaw conversion.Hub) error {
 	dst := dstRaw.(*v1alpha2.MySQLAADUser)
 
 	// ObjectMeta
 	dst.ObjectMeta = src.ObjectMeta
 
+	// If there are stashed values in the annotations then we need to
+	// retrieve them first.
+	var stashedValues stashedMySQLAADUserFields
+	found, err := getStashedAnnotation(src.ObjectMeta, &stashedValues)
+	if err != nil {
+		return err
+	}
+	if found {
+		dst.Spec.Roles = stashedValues.Roles
+		dst.Spec.DatabaseRoles = stashedValues.DatabaseRoles
+
+		// Clear out the annotation to avoid confusion.
+		clearStashedAnnotation(&dst.ObjectMeta)
+	}
+
 	// Spec
 	dst.Spec.ResourceGroup = src.Spec.ResourceGroup
 	dst.Spec.Server = src.Spec.Server
 	dst.Spec.AADID = src.Spec.AADID
 	dst.Spec.Username = src.Spec.Username
 
-	// v1alpha1 doesn't support server-level roles, only
-	// database-level ones, so we move them into the new field.
-	dst.Spec.Roles = []string{}
-	dst.Spec.DatabaseRoles = make(map[string][]string)
+	if dst.Spec.Roles == nil {
+		dst.Spec.Roles = []string{}
+	}
+	if dst.Spec.DatabaseRoles == nil {
+		dst.Spec.DatabaseRoles = make(map[string][]string)
+	}
 	dst.Spec.DatabaseRoles[src.Spec.DBName] = append([]string(nil), src.Spec.Roles...)
```

Comment:
Is this just a no-op in the case where there was a stashed annotation? (The assignment has actually already happened but we're just doing it again?)

Reply:
Not necessarily a no-op- if a user was doing an edit they might send back a resource with changed roles for the selected database (but the original values in the annotation). But in general probably.